### PR TITLE
Canary upgrade status in Control Plane card

### LINKED ISSUE #5524: Indicate the status of a canary upgrade in the control plane card
Istio's recommended upgrade method is the "Canary upgrade".
Docs: https://istio.io/latest/docs/setup/upgrade/canary/

If possible, Kiali should detect and indicate about an ongoing canary upgrade. If it is not possible to detect, there is already some configuration that is used to support moving/migrating workloads for canary upgrades.

Under the assumption that the new control plane will be deployed on the same namespace as the old one, the "Control Plane" card is probably a good place to put any indicators.

A few data that is worth to show:
* Since there will be two active control planes, show simple indication bout pod/deployment status if it's ready (pod status?)
* The percentage of workloads that still in the stable version and the percentage of workloads that already migrated

Epic #4435

### PR DESCRIPTION
**Pull request by leandroberetta**
Resolves #5524

WIP.

image

### CODE REVIEW

**Comment by leandroberetta** [comment]
Some updates after UX meeting:

image

**Comment by jshaughn** [comment]
Those changes look good.  Are there other changes pending?  We had talked about somehow indicating which namespaces had migrated, the blank space to the bottom left (not sure we had a solution) and, minor, sentence-casing the title (e.g. `Canary Upgrade Status` -> `Canary upgrade status`).

**Comment by leandroberetta** [comment]
> 

Yes, I have some of that changes pending. I will work on that today.

**Comment by josunect** [comment]
I like the new donuts chart! It looks nice. 

- When I tested, I see a difference with the screenshot: 

image

I cannot see the total number of namespaces - But they are in the api response: 
`{"migratedNamespaces":["bookinfo"],"pendingNamespaces":["alpha","beta","sleep"]}`

But, maybe these changes were not updated yet?

- Another thing is that if I don't set revision=canary I can still see the canary grid: 

image

For what I understood from the documentation https://istio.io/latest/docs/setup/upgrade/canary/, if the revision is not set to canary we won't have the canary control plane, so maybe in this case it would make sense to hide the graph?

**Comment by leandroberetta** [comment]
@jshaughn @josunect  the following represents the changes spoke in the meeting:

image

**Comment by jshaughn** [comment]
@leandroberetta For me everything seems to be addressed and looks good.  Maybe @andrew-ronaldson can look at the screenshot above and comment.  One very minor nit (and not related to PR), we may want to change `Istio Config` to `Istio config` for the consistent casing.

**Comment by leandroberetta** [comment]
Thanks @jshaughn, yes, and maybe also the other ones too like Outbound Policy (Outbound policy?).

**Comment by josunect** [comment]
> > I like the new donuts chart! It looks nice.
> > 
> > * When I tested, I see a difference with the screenshot:
> > 
> > image
> > I cannot see the total number of namespaces - But they are in the api response: `{"migratedNamespaces":["bookinfo"],"pendingNamespaces":["alpha","beta","sleep"]}`
> > But, maybe these changes were not updated yet?
> > 
> > * Another thing is that if I don't set revision=canary I can still see the canary grid:
> > 
> > image
> > For what I understood from the documentation https://istio.io/latest/docs/setup/upgrade/canary/, if the revision is not set to canary we won't have the canary control plane, so maybe in this case it would make sense to hide the graph?
> 
> Yes, sorry, I didn't pushed the latest code as I have things to do yet, but thanks a lot for the preliminar review.
> 
> The canary status apperars when in Kiali CR you configured the current and upgrade versions as well the feature flag for the canaries. Then, if non of the namespaces have the label for the canary, the progress is ok to be 0.

Thanks, @leandroberetta ! I have reviewed the new changes and the new dashboard looks great. 

So I guess we should just wait for the review of Andrew and the comments from @jshaughn - Maybe the casing changes can be also be applied to the "Min TLS Version" - "Min TLS version"

**Comment by leandroberetta** [comment]
@josunect yes, I think we should wait for UX feedback and also I'm working on the Cypress testing. 

Thanks for the review.

**Comment by andrew-ronaldson** [comment]
> image

Late response* Seeing this screenshot I think the details section looks better aligned top.
For PF consistency anywhere we use title casing (Every word capitalized) it should appear Sentence case.
Istio Config, Outbound Policy, Min TLS version.

Thanks for sharing this!

**Comment by leandroberetta** [comment]
@jshaughn @josunect this issue is ready for final review.

Regarding to Cypress testing, I added the test that check that if there is no canary configured, the section would not be present (this is for Maistra, as they don't support canaries). I'm looking for a way to modify the environment to install another control plane and re configure Kiali with the canary information (the new control plane), for now it's just researching ways. Do you prefer wait for the research results or should we merged this (covering Maistra part) and continue this in a follow up PR?. I think I prefer the second one to stop rebasing in case of further updates to the cp card.

File: business/mesh.go [hunk 1]
```diff
@@ -642,3 +642,50 @@ func (in *MeshService) IstiodResourceThresholds() (*models.IstiodThresholds, err
...
+	pendingNss, err = in.k8s.GetNamespaces(fmt.Sprintf("istio.io/rev=%s", current))
+	if err != nil {
+		return nil, err
+	}
+	for _, ns := range pendingNss {
+		pendingNsList = append(pendingNsList, ns.Name)
+	}
+
+	status := &models.CanaryUpgradeStatus{
+		CurrentVersion:     conf.ExternalServices.Istio.IstioCanaryRevision.Current,
```
Comment (josunect) [inline business/mesh.go:684]: // It looks like this value is already assigned so it probably can be accessed as: 

CurrentVersion:     current

**Review by josunect — COMMENTED** [review]
Tested and it looks great! 
I've just added some minor comments. 

Thanks!

File: frontend/src/pages/Overview/ControlPlaneVersionBadge.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+type Props = {
+    version: string;
+    isCanary: boolean;
+};
+
+class ControlPlaneVersionBadge extends React.Component<Props> {
+    render() {
+        return (
+            <>
+              <Label style={{ marginLeft: 5 }} color={this.props.isCanary ? "blue": "red"} isCompact>{this.props.version}</Label>
```
[the remark this answers is not in the document]
Comment (leandroberetta) [reply frontend/src/pages/Overview/ControlPlaneVersionBadge.tsx]: Yes! I forgot to push :)

**Review by josunect — APPROVED** [review]
LGTM!

**Pull request merged**